### ISSUE TITLE
How to tell esbuild to bundle more CSS from a JS loader?

### ISSUE THREAD

**Issue by zaydek** [evanw/esbuild#1138]
I’m experimenting with _Sass Components_. The idea is that I can author Sass components in my JS file and use esbuild to do two things: a) use a plugin to compile Sass code down to CSS and b) return a JS object that associates raw Sass strings to React components with generated class names.

Here’s the user story:

```js
import sass from "sass-components"

const MyComponent = sass.div(`
  display: flex;
  flex-direction: row;
  justify-content: center;
  // ...
`)

export default function App() {
  return (
    <MyComponent>
      {/* ... */}
    </MyComponent>
  )
}
```

I think what I need is the ability to tell esbuild to return two `onLoadResult`s for one plugin. So I want to be able to return:

```js
build.onResolve({ namespace: "sass-components-ns" }, args => {
  // ...
  {
    contents: /* Generated CSS code */,
    loader: "css",
  }
})
```

and the JavaScript object:

```js
build.onResolve({ namespace: "sass-components-ns" }, args => {
  // ...
  return {
    contents: `
      export default {
        ["display: flex; ..."]: props => <div className={\`sass-${hash}\`} {...props} />,
      }
    `,
    loader: "jsx",
  }
})
```

In the same plugin. Is there way I can resolve two loaders, one for bundling CSS (with the rest of the bundled CSS) and one for the generated JavaScript object?

I have the JS `onLoadResult` working fine, so I just need a way to tell esbuild to _also_ bundle the CSS generated from this plugin. Because esbuild is already generating a bundled CSS file, I want to include / bundle the generated CSS from this plugin as well, but I don’t know how to tell esbuild to do both of these things from inside of one plugin.

Thank you in advance.

**Comment by zaydek** [comment]
Oh, maybe I can register two plugins like this:

```js
build.onLoad({ filter: /.*/, namespace: "sass-components-ns" }, async args => { /* Handle JS */ }
build.onLoad({ filter: /.*/, namespace: "sass-components-ns" }, async args => { /* Handle CSS */ }
```

This is assuming esbuild will let me handle the same filter or namespace more than once, which I assume it will. 🤔

**Comment by evanw** [comment]
The way to do this is to import the CSS in your JS, just like you would if you were writing the code manually. You can either do this by importing from a special import path like `<css>` (or even `<css:...>`) and using another resolve + load to handle it, or you could maybe instead import from a data URL directly to avoid having to do the extra round-trips:

```diff
   return {
     contents: `
+      import "data:text/css,/* Generated CSS code */"
       export default {
         ["display: flex; ..."]: props => <div className={\`sass-${hash}\`} {...props} />,
       }
     `,
     loader: "jsx",
   }
```

**Comment by zaydek** [comment]
Thank you! This is really cool.

CSS-in-JS libraries always have their own runtime or use Babel to do ahead-of-time compilation, so I like this because it’s zero-runtime and you don’t need Babel. It’s just install this plugin and use Sass inside of your JS with managed class names.

If I get it working I’ll share the MVP repo for you and others to look at. Should help with some of the users in #20 with similar use cases.

**Comment by zaydek** [comment]
Not sure I’ll actually use this but I ended up making a MVP prototype here: https://github.com/zaydek/retro-test-sass-components/blob/master/src/App.js. Tiny real-world demo here: https://optimistic-bartik-6e35f2.netlify.app.

@evanw The inline tip worked out really well. I just needed to encode the result but once I did that it was working.

```js
return {
  contents: `
    import "data:text/css,${encodeURI(result.css.toString())}"
    ${contents}
  `,
  loader: "jsx",
}
```

I was a little worried that the bundled output would include the whole string as you use a comment to delimit where a plugin our source file originates, but I was very glad to see you truncate after 80 or so characters:

```css
/* <data:text/css,button%20%7B%0A%20%20border:%20unset;%0A%20%20back...> */
button {
  border: unset;
  background-color: unset;
}
```

Anyway, once I coded this up, I realized the fatal flaw that I can’t interpolate class names using `${syntax}` because everything is static.

Anyway, virtual high-five!
